fem/qfunction.cpp

```diff
@@ -0,0 +1,265 @@
+
+   qspace = new QuadratureSpace(mesh, in);
+   own_qspace = true;
+
+   in >> ident; MFEM_VERIFY(ident == "VDim:", msg);
+   in >> vdim;
+
+   Load(in, vdim*qspace->GetSize());
+}
+
+void QuadratureFunction::SetSpace(QuadratureSpaceBase *qspace_, int vdim_)
+{
+   if (qspace_ != qspace)
+   {
+      if (own_qspace) { delete qspace; }
+      qspace = qspace_;
+      own_qspace = false;
+   }
+   vdim = (vdim_ < 0) ? vdim : vdim_;
+   SetSize(vdim*qspace->GetSize());
+}
+
+void QuadratureFunction::SetSpace(
+   QuadratureSpaceBase *qspace_, double *qf_data, int vdim_)
+{
+   if (qspace_ != qspace)
+   {
+      if (own_qspace) { delete qspace; }
+      qspace = qspace_;
+      own_qspace = false;
+   }
+   vdim = (vdim_ < 0) ? vdim : vdim_;
+   NewDataAndSize(qf_data, vdim*qspace->GetSize());
+}
+
+void QuadratureFunction::Save(std::ostream &os) const
+{
+   GetSpace()->Save(os);
+   os << "VDim: " << vdim << '\n'
+      << '\n';
+   Vector::Print(os, vdim);
+   os.flush();
+}
+
+void QuadratureFunction::ProjectGridFunction(const GridFunction &gf)
+{
+   SetVDim(gf.VectorDim());
+
+   if (auto *qs_elem = dynamic_cast<QuadratureSpace*>(qspace))
+   {
+      const FiniteElementSpace &gf_fes = *gf.FESpace();
+      const bool use_tensor_products = UsesTensorBasis(gf_fes);
+      const ElementDofOrdering ordering = use_tensor_products ?
+                                          ElementDofOrdering::LEXICOGRAPHIC :
+                                          ElementDofOrdering::NATIVE;
+
+      // Use element restriction to go from L-vector to E-vector
+      const Operator *R = gf_fes.GetElementRestriction(ordering);
+      Vector e_vec(R->Height());
+      R->Mult(gf, e_vec);
+
+      // Use quadrature interpolator to go from E-vector to Q-vector
+      const QuadratureInterpolator *qi = gf_fes.GetQuadratureInterpolator(*qs_elem);
+      qi->SetOutputLayout(QVectorLayout::byVDIM);
+      qi->DisableTensorProducts(!use_tensor_products);
+      qi->Values(e_vec, *this);
+   }
+   else if (auto *qs_face = dynamic_cast<FaceQuadratureSpace*>(qspace))
+   {
+      const FiniteElementSpace &gf_fes = *gf.FESpace();
+      const bool use_tensor_products = UsesTensorBasis(gf_fes);
+      const ElementDofOrdering ordering = use_tensor_products ?
+                                          ElementDofOrdering::LEXICOGRAPHIC :
+                                          ElementDofOrdering::NATIVE;
+
+      const FaceType face_type = qs_face->GetFaceType();
+
+      // Use element restriction to go from L-vector to E-vector
+      const Operator *R = gf_fes.GetFaceRestriction(
+                             ordering, face_type, L2FaceValues::SingleValued);
+      Vector e_vec(R->Height());
+      R->Mult(gf, e_vec);
+
+      // Use quadrature interpolator to go from E-vector to Q-vector
+      const FaceQuadratureInterpolator *qi =
+         gf_fes.GetFaceQuadratureInterpolator(qspace->GetIntRule(0), face_type);
+      qi->SetOutputLayout(QVectorLayout::byVDIM);
```

Comment:
Looks like this addresses issue https://github.com/mfem/mfem/issues/2865 ?

Reply:
Yes, I think so. I forgot about that issue, I hope this fixes it (and if not it should be straightforward to adjust it)